thiduzz/terraform-ecs-fargate-boilerplate
# Add ECR Repositories and permission policies to specific users

### Proposal

Given that we want to have different levels of access to ECR repositories and not every repository should be available with the same of access, the idea is to create 2 different users, one with pull access only and another with pull and push. This users should only be able to operate in a set of repositories (not application images, only base images). This can be extended in the future to allow cross-account pulls (between staging/prod).

Scenarios considered:

- [x] Logged out
- [x] Full access - pull
- [x] Full access - push
- [x] Full access - Non base image
- [x] Pull access - pull
- [x] Pull access - push
- [x] Pull access - Non base image

### Basic Testing Instructions

1. Apply the changes: `terraform apply`
2. Retrieve the credentials generates (sensitive) and configure your AWS CLI
3. Login with Docker: `aws --profile {YOUR_PROFILE} ecr get-login-password --region {YOUR_REGION} | docker login --username AWS --password-stdin {YOUR_ACCOUNT}.dkr.ecr.{YOUR_REGION}.amazonaws.com`
4. Build an image and attempt the following wiith the repos `base-php7.4-alpine`/`base-php8.0-alpine`
4.1. Attempt with the full access user pushing/pulling: ie. `docker pull 411705946934.dkr.ecr.eu-central-1.amazonaws.com/base-php7.4-alpine:1.1` or `docker push 411705946934.dkr.ecr.eu-central-1.amazonaws.com/base-php7.4-alpine:1.1` (should succeed)
4.2 Attempt with the pull access user pulling: ie. `docker pull 411705946934.dkr.ecr.eu-central-1.amazonaws.com/base-php7.4-alpine:1.1` (should succeed)
4.3 Attempt with the pull access user pushing: ie. `docker push 411705946934.dkr.ecr.eu-central-1.amazonaws.com/base-php7.4-alpine:1.1` (should fail)
5. Attempt pulling or pushing the an app repository image (any user) (should fail)
